Pt 154245674 oracle api

hanssv opened the pull request:
WebSocket API and life cycle example using the API

hanssv commented:
interaction_id and query_id are used in a confusing way - will try to make a patch for this...

knarz commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+example, but it should be straightforward to split the example into one node
+running the oracle and other nodes querying that oracle. In the default node
+configuration the websocket interface is at *port 3014* - i.e. connecting to
+`ws://127.0.0.1:3114/websocket` open a websocket connection.
```
s/open/opens/

knarz commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+
+On a freshly started node, connect a websocket. (Here we use
+[WSCat](https://github.com/websockets/wscat) for the websocket connection.)
+Upon connection the client automatically subscribe to the _mined_block_-event,
```
s/subscribe/subscribes/

knarz commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+On a freshly started node, connect a websocket. (Here we use
+[WSCat](https://github.com/websockets/wscat) for the websocket connection.)
+Upon connection the client automatically subscribe to the _mined_block_-event,
+i.e. everytime the node mine a new block an event is pushed to the client. We
```
s/mine/mines/

knarz commented on oracles/oracle_ws_api.md:
````diff
@@ -0,0 +1,427 @@
+< {"action":"register","origin":"oracle","payload":{"result":"ok","oracle_id":"ok$3jzZyCLFtHVD7yVdEhGJFM3LjeXrKqWxnHbCYzhnrrR4DkdFtaJuxQvrR8VbbXExDPkCHFAei5q969JA6EayQpb8z5C3Mf"}}
+```
+
+The register transaction use the same format for the payload as the HTTP API
````
s/use/uses/

knarz commented on oracles/oracle_ws_api.md:
````diff
@@ -0,0 +1,427 @@
+```
+
+We subscribe for `"type":"query"`, i.e. we want to get notified when the oracle
+with the given id receive a query. Of course we want to test this, so let us
````
s/receive/receives/

knarz commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+per 1000 blocks (just a preliminary value to test the concept). I.e. the cost
+for our test transaction is 7.
+
+Not much will happen until the node mine another block (and the query
```
s/mine/mines/

knarz commented on oracles/oracle_ws_api.md:
````diff
@@ -0,0 +1,427 @@
+```
+
+We subscribe for `"type":"response"`, i.e. we want to get notified when the
+oracle answer the query with the given id. Of course we want to test this as
````
s/answer/answers/

knarz commented on oracles/oracle_ws_api.md:
````diff
@@ -0,0 +1,427 @@
+< {"action":"new_oracle_response","origin":"node","payload":{"interaction_id":"oi$4RZoMEkm8QuuhJiiq53dd5pE4VstCthYRjBHgUKdhAhe7rLEr","response":"I am fine, thanks!"}}
+```
+
+This conclude the walk through of the oracle life cycle example, which also
````
s/conclude the walk through/concludes the walkthrough/

knarz commented on oracles/oracle_ws_api.md:
````diff
@@ -0,0 +1,427 @@
+```
+
+This conclude the walk through of the oracle life cycle example, which also
+exercise the websocket API for Oracles fully.
````
s/exercise/exercises/

lucafavatella commented on oracles/oracles.md:
```diff
@@ -16,6 +16,8 @@
 
 ## [Oracle life cycle examples](./oracle_life_cycle.md)
 
+## [Oracle WebSocket API](./oracle_ws_api.md)
```
When we spoke today I understood we wanted to keep epoch API in this repo though segregated somehow. I would propose last section in this document `## epoch node API` and separate folder `./epoch_api` containing API docs.

lucafavatella commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+(all signing is done by a node - and that node is associated with exactly one
+account) it is somewhat limited what you can do with oracles. I.e. in practice
+you need a node per account - obviously this is a limitation we should try to
+lift as soon as possible.
```
This looks an action. Is this tracked somewhere and discussed?

knarz replied:
I created this task a while ago, which is related: https://www.pivotaltracker.com/story/show/153451598

lucafavatella commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+[back](./oracles.md)
```
May I suggest splitting this document in two parts:
* A document that acts as a surrogate for the Swagger schema for websocket
* A document that documents to intended usage of API (the portion "For simplicity, we only work with a single node...")

lucafavatella commented on oracles/oracle_ws_api.md:
```diff
@@ -0,0 +1,427 @@
+  | Name | Type | Description | Required |
+  | ---- | ---- | ----------- | -------- |
+  | result | string | `ok` or error reason | Yes |
+  | query_id | string | Hash for created query | Yes on success |
```
Is the query_id an id per-oracle or across all oracles?

lucafavatella replied:
and is user meant to be able to extract oracle from query (I think not needed)?

hanssv replied:
query_id is unique (across all oracles) 

no, you can't extract the oracle_id from the query_id

hanssv commented:
Introduced epoch_api and put the more formal description of the Oracle WS API there. Kept the _intended usage_ part under Oracles...

Also updated the State tree description for Oracles.

lucafavatella commented on oracles/oracle_state_tree.md:
```diff
@@ -58,16 +60,13 @@ has the benefits:
-- Should the oracle state tree be a separate tree or
-  be joined with the accounts state tree?
-- Define the contents of the oracle interaction object.
+**NOTE:** The refunding has not yet been implemented
```
Thanks for the note.

hanssv replied:
It is also tracked by PT-#154292757
